packages/axelar-wasm-std/src/msg_id/tx_hash.rs

```diff
@@ -0,0 +1,128 @@
+    }
+
+    pub fn new(tx_id: impl Into<[u8; 32]>) -> Self {
+        Self {
+            tx_hash: tx_id.into(),
+        }
+    }
+}
+
+const PATTERN: &str = "^[0-9a-f]{64}$";
+lazy_static! {
+    static ref REGEX: Regex = Regex::new(PATTERN).expect("invalid regex");
+}
+
+impl FromStr for HexTxHash {
+    type Err = Report<Error>;
+
+    fn from_str(message_id: &str) -> Result<Self, Self::Err>
+    where
+        Self: Sized,
+    {
+        // the PATTERN has exactly two capture groups, so the groups can be extracted safely
+        REGEX
+            .captures(message_id)
+            .ok_or(Error::InvalidMessageID {
+                id: message_id.to_string(),
+                expected_format: PATTERN.to_string(),
+            })?;
+        Ok(HexTxHash {
+            tx_hash: HexBinary::from_hex(message_id)
+                .change_context(Error::InvalidTxHash(message_id.to_string()))?
+                .as_slice()
+                .try_into()
+                .map_err(|_| Error::InvalidTxHash(message_id.to_string()))?,
+        })
+    }
+}
+
+impl Display for HexTxHash {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            f,
+            "{}",
+            HexBinary::from(self.tx_hash).to_hex(),
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+
+    use super::*;
+
+    fn random_hash() -> String {
+        let mut bytes = vec![];
+        for _ in 0..32 {
+            let byte: u8 = rand::random();
+            bytes.push(byte)
+        }
+        format!("{}", HexBinary::from(bytes).to_hex())
+    }
+
+    #[test]
+    fn should_parse_msg_id() {
+        let res = HexTxHash::from_str(
+            "7cedbb3799cd99636045c84c5c55aef8a138f107ac8ba53a08cad1070ba4385b",
+        );
+        assert!(res.is_ok());
+
+        for _ in 0..1000 {
+            let msg_id = random_hash();
+
+            let res = HexTxHash::from_str(&msg_id);
+            let parsed = res.unwrap();
+            assert_eq!(parsed.tx_hash_as_hex(), msg_id.clone().try_into().unwrap());
+            assert_eq!(parsed.to_string(), msg_id);
+        }
+    }
+
+    #[test]
+    fn should_not_parse_msg_id_with_wrong_length_tx_hash() {
+        let tx_hash = random_hash();
+        // too long
+        let res = HexTxHash::from_str(&format!("{}ff", tx_hash));
+        assert!(res.is_err());
+
+        // too short
+        let res =
+            HexTxHash::from_str(&format!("{}", &tx_hash[..tx_hash.len() - 2]));
+        assert!(res.is_err());
+    }
+
+    #[test]
+    fn should_not_parse_msg_id_with_uppercase_tx_hash() {
+        let tx_hash = &random_hash();
+        let res = HexTxHash::from_str(&tx_hash.to_uppercase());
+        assert!(res.is_err());
+    }
+
+    #[test]
+    fn should_not_parse_msg_id_with_non_hex_tx_hash() {
+        let msg_id = "82GKYvWv5EKm7jnYksHoh3u5M2RxHN2boPreM8Df4ej9";
+        let res = HexTxHash::from_str(msg_id);
+        assert!(res.is_err());
+    }
+}
```

Comment:
```suggestion
    #[test]
    fn should_not_parse_msg_id_with_0x_prefix() {
        let msg_id = "0x7cedbb3799cd99636045c84c5c55aef8a138f107ac8ba53a08cad1070ba4385b-1";
        let res = HexTxHash::from_str(msg_id);
        assert!(res.is_err());
    }

    #[test]
    fn should_not_parse_msg_id_with_event_index() {
        let tx_hash = random_hash();
        let res = HexTxHash::from_str(&format!("{}-1", tx_hash));
        assert!(res.is_err());
    }
}
```

Just copy-pasted from our duplicate PR: https://github.com/axelarnetwork/axelar-amplifier/pull/532/files#diff-21e519f3544fb0dc94d4d87daef8ee69a1bd012bc43723dd93417cf95284680bR128-R141.

Reply:
Done